modules/apps/portal/portal-util-test/src/testIntegration/java/com/liferay/portal/util/test/PortalImplWidgetURLResponseTest.java

```diff
@@ -0,0 +1,126 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.util.test;
+
+import com.liferay.arquillian.extension.junit.bridge.junit.Arquillian;
+import com.liferay.journal.constants.JournalContentPortletKeys;
+import com.liferay.layout.test.util.ContentLayoutTestUtil;
+import com.liferay.layout.test.util.LayoutTestUtil;
+import com.liferay.portal.kernel.model.Company;
+import com.liferay.portal.kernel.model.Group;
+import com.liferay.portal.kernel.model.GroupConstants;
+import com.liferay.portal.kernel.model.Layout;
+import com.liferay.portal.kernel.model.Portlet;
+import com.liferay.portal.kernel.model.User;
+import com.liferay.portal.kernel.service.CompanyLocalService;
+import com.liferay.portal.kernel.service.GroupLocalService;
+import com.liferay.portal.kernel.service.LayoutLocalService;
+import com.liferay.portal.kernel.service.PortletLocalServiceUtil;
+import com.liferay.portal.kernel.test.rule.AggregateTestRule;
+import com.liferay.portal.kernel.test.util.TestPropsValues;
+import com.liferay.portal.kernel.test.util.UserTestUtil;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.util.Portal;
+import com.liferay.portal.test.rule.Inject;
+import com.liferay.portal.test.rule.LiferayIntegrationTestRule;
+
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ * @author István András Dézsi
+ */
+@RunWith(Arquillian.class)
+public class PortalImplWidgetURLResponseTest {
+
+	@ClassRule
+	@Rule
+	public static final AggregateTestRule aggregateTestRule =
+		new LiferayIntegrationTestRule();
+
+	@Before
+	public void setUp() throws Exception {
+		_company = _companyLocalService.getCompany(
+			TestPropsValues.getCompanyId());
+
+		_group = _groupLocalService.fetchGroup(
+			_company.getCompanyId(), GroupConstants.GUEST);
+
+		_layout = _layoutLocalService.fetchDefaultLayout(
+			_group.getGroupId(), false);
+
+		_user = TestPropsValues.getUser();
+
+		UserTestUtil.setUser(_user);
+	}
+
+	@Test
+	public void testWidgetURLResponse() throws Exception {
+		Layout draftLayout = _layout.fetchDraftLayout();
+
+		ThemeDisplay themeDisplay = ContentLayoutTestUtil.getThemeDisplay(
+			_company, _group, draftLayout);
+
+		themeDisplay.setPortalURL(_company.getPortalURL(_group.getGroupId()));
+
+		String portletId = LayoutTestUtil.addPortletToLayout(
+			draftLayout, JournalContentPortletKeys.JOURNAL_CONTENT);
+
+		ContentLayoutTestUtil.publishLayout(draftLayout, _layout);
+
+		Portlet portlet = PortletLocalServiceUtil.getPortletById(
+			_company.getCompanyId(), portletId);
+
+		URL widgetURL = new URL(_portal.getWidgetURL(portlet, themeDisplay));
```

Comment:
@IstvanD According to the steps in your ticket 

> Create a new public page and add the fragment to it, then publish it
> Log out and visit the page as guest user
You should not be able to see the same themeDisplay instance you used to publish layout, while in your test, you are doing `URL widgetURL = new URL(_portal.getWidgetURL(portlet, themeDisplay));` using the themeDisplay created with draftLayout.

This does not make sense to me.

Reply:
@tinatian I pushed the changes.